File: orders-bs.js
/**
 * Beauty Star Orders - Server v4
 * Fixes: Windows Auth, TCP port, named instance
 */
const http = require('http');
const url  = require('url');
const sql  = require('mssql');
const fs   = require('fs');
const path = require('path');

const PORT    = 3001;
const LOG_CSV = path.join(__dirname, 'error-log.csv');

function writeLog(type, store, message){
  if(!fs.existsSync(LOG_CSV))
    fs.writeFileSync(LOG_CSV, 'Date,Time,Type,Store,Message\n');
  const now  = new Date();
  const safe = s => `"${String(s).replace(/"/g,'""')}"`;
  fs.appendFileSync(LOG_CSV,
    `${safe(now.toLocaleDateString('en-GB'))},${safe(now.toLocaleTimeString('en-GB'))},${safe(type)},${safe(store||'')},${safe(message)}\n`
  );
  console.log(`[${type}] ${store||''} ${message}`);
}

// ── TRY MULTIPLE CONNECTION CONFIGS ───────────────────────────────────────
const CONFIGS = [
  // Works on ANY branch PC — localhost so same files work everywhere
  { server:'localhost', options:{ instanceName:'SQLEXPRESS', trustServerCertificate:true, encrypt:false }, user:'beautystar', password:'Bs2024!orders' },
  { server:'127.0.0.1', options:{ instanceName:'SQLEXPRESS', trustServerCertificate:true, encrypt:false }, user:'beautystar', password:'Bs2024!orders' },
  { server:'localhost\SQLEXPRESS', options:{ trustServerCertificate:true, encrypt:false }, user:'beautystar', password:'Bs2024!orders' },
  { server:'localhost', options:{ instanceName:'SQLEXPRESS', trustedConnection:true, trustServerCertificate:true, encrypt:false } },
  { server:'localhost\SQLEXPRESS', options:{ trustedConnection:true, trustServerCertificate:true, encrypt:false } },
];

const BASE = { database:'ITSDryStock', pool:{ max:5, min:0, idleTimeoutMillis:30000 } };

let pool = null;
let connectedConfig = null;

async function getPool(){
  if(pool) return pool;
  for(const cfg of CONFIGS){
    const full = { ...BASE, ...cfg, options:{ ...BASE.options, ...cfg.options } };
    const label = `${full.server}${full.options.instanceName?'\\'+full.options.instanceName:''}:${full.port||'auto'}`;
    try{
      writeLog('CONNECT','',`Trying ${label}...`);
      pool = await sql.connect(full);
      connectedConfig = label;
      writeLog('CONNECT','',`SUCCESS via ${label}`);
      return pool;
    }catch(e){
      writeLog('CONNECT_FAIL','',`${label}: ${e.message.split('\n')[0]}`);
      pool = null;
      try{ await sql.close(); }catch(x){}
    }
  }
  throw new Error('Could not connect — see error-log.csv for details');
}

const STOCK_SQL = `
  SELECT spr.spr_stkcode AS code, bp.bxp_instock AS qty
  FROM   BranchProducts bp
  JOIN   Products          p   ON p.prd_pk      = bp.bxp_prdfk
  JOIN   SupplierProducts  spr ON spr.spr_prdfk  = bp.bxp_prdfk AND spr.spr_main=1 AND spr.spr_inactive=0
  JOIN   Branches          br  ON br.brn_pk     = bp.bxp_brnfk
  WHERE  br.brn_name = @store AND bp.bxp_inactive=0 AND p.prd_deleted=0
    AND  spr.spr_stkcode IS NOT NULL AND spr.spr_stkcode != ''`;

const BRANCHES_SQL = `SELECT brn_pk, brn_number, brn_name FROM Branches ORDER BY brn_name`;

const ALL_SQL = `
  SELECT br.brn_name AS store, spr.spr_stkcode AS code, bp.bxp_instock AS qty
  FROM   BranchProducts bp
  JOIN   Products          p   ON p.prd_pk      = bp.bxp_prdfk
  JOIN   SupplierProducts  spr ON spr.spr_prdfk  = bp.bxp_prdfk AND spr.spr_main=1 AND spr.spr_inactive=0
  JOIN   Branches          br  ON br.brn_pk     = bp.bxp_brnfk
  WHERE  bp.bxp_inactive=0 AND p.prd_deleted=0
    AND  spr.spr_stkcode IS NOT NULL AND spr.spr_stkcode != ''`;

const httpServer = http.createServer(async (req, res) => {
  res.setHeader('Access-Control-Allow-Origin','*');
  res.setHeader('Access-Control-Allow-Methods','GET,OPTIONS');
  res.setHeader('Content-Type','application/json');
  if(req.method==='OPTIONS'){ res.writeHead(204); res.end(); return; }

  const p = url.parse(req.url, true);

  if(p.pathname==='/health'){
    let dbOk=false;
    try{ await getPool(); dbOk=true; }catch(e){}
    res.writeHead(200);
    res.end(JSON.stringify({ status:'ok', db:dbOk?'connected':'error', via:connectedConfig||'none', time:new Date().toISOString() }));
    return;
  }

  if(p.pathname==='/branches'){
    try{
      const db=await getPool();
      const r=await db.request().query(BRANCHES_SQL);
      writeLog('INFO','',`${r.recordset.length} branches returned`);
      res.writeHead(200); res.end(JSON.stringify(r.recordset));
    }catch(e){ pool=null; writeLog('ERROR','branches',e.message); res.writeHead(500); res.end(JSON.stringify({error:e.message})); }
    return;
  }

  if(p.pathname==='/stock'){
    const store=p.query.store;
    if(!store){ res.writeHead(400); res.end(JSON.stringify({error:'Missing ?store='})); return; }
    try{
      const db=await getPool();
      let r1=db.request(); r1.input('store',sql.NVarChar,store);
      let rows=(await r1.query(STOCK_SQL)).recordset;
      if(!rows.length){
        let r2=db.request(); r2.input('store',sql.NVarChar,`%${store}%`);
        rows=(await r2.query(STOCK_SQL.replace('br.brn_name = @store','br.brn_name LIKE @store'))).recordset;
      }
      const map={};
      rows.forEach(r=>{ if(r.code) map[r.code.trim()]=parseFloat(r.qty)||0; });
      writeLog('STOCK',store,`${rows.length} products returned`);
      res.writeHead(200); res.end(JSON.stringify(map));
    }catch(e){ pool=null; writeLog('ERROR',store,e.message); res.writeHead(500); res.end(JSON.stringify({error:e.message})); }
    return;
  }

  if(p.pathname==='/stock/all'){
    try{
      const db=await getPool();
      const rows=(await db.request().query(ALL_SQL)).recordset;
      const out={};
      rows.forEach(r=>{ if(!out[r.store])out[r.store]={}; out[r.store][r.code.trim()]=parseFloat(r.qty)||0; });
      writeLog('STOCK','ALL',`${rows.length} rows, ${Object.keys(out).length} stores`);
      res.writeHead(200); res.end(JSON.stringify(out));
    }catch(e){ pool=null; writeLog('ERROR','all',e.message); res.writeHead(500); res.end(JSON.stringify({error:e.message})); }
    return;
  }

  if(p.pathname==='/log'){
    if(!fs.existsSync(LOG_CSV)) fs.writeFileSync(LOG_CSV,'Date,Time,Type,Store,Message\n');
    res.setHeader('Content-Type','text/csv');
    res.setHeader('Content-Disposition','attachment; filename="error-log.csv"');
    res.writeHead(200); res.end(fs.readFileSync(LOG_CSV,'utf8'));
    return;
  }

  res.writeHead(404); res.end(JSON.stringify({error:'Try /health /branches /stock?store=Name /stock/all /log'}));
});

httpServer.listen(PORT,'0.0.0.0',()=>{
  writeLog('START','',`Server on port ${PORT}`);
  console.log(`\n✅  Beauty Star Orders server running`);
  console.log(`    http://127.0.0.1:${PORT}/health`);
  console.log(`    http://127.0.0.1:${PORT}/branches`);
  console.log(`    http://127.0.0.1:${PORT}/stock?store=Chapeltown`);
  console.log(`    error-log.csv: ${LOG_CSV}\n`);
  console.log(`[*] KEEP THIS WINDOW OPEN\n`);
});

httpServer.on('error',e=>{
  if(e.code==='EADDRINUSE'){ writeLog('ERROR','','Port '+PORT+' already in use'); console.log('\n[!] Port already in use — close the other window\n'); }
  else writeLog('ERROR','','Server: '+e.message);
  process.exit(1);
});

process.on('SIGINT',async()=>{ writeLog('STOP','','Shutdown'); if(pool)await pool.close(); process.exit(0); });

File: orders-bs.py
"""
Beauty Star Orders - Server (Python fallback)
Run this if Node.js is not installed.
Requires: pip install pyodbc
"""
import http.server, json, urllib.parse, sys
from datetime import datetime
try:
    import pyodbc
except ImportError:
    print("Run: pip install pyodbc"); sys.exit(1)

SERVER   = r'.\SQLEXPRESS'
DATABASE = 'ITSDryStock'
PORT     = 3001
DRIVERS  = ['ODBC Driver 17 for SQL Server','ODBC Driver 18 for SQL Server','SQL Server']

STOCK_SQL = """
  SELECT bar.bar_barcode AS code, bp.bxp_instock AS qty
  FROM   BranchProducts bp
  JOIN   Products   p   ON p.prd_pk    = bp.bxp_prdfk
  JOIN   Barcodes   bar ON bar.bar_prdfk = bp.bxp_prdfk AND bar.bar_default = 1
  JOIN   Branches   br  ON br.brn_pk   = bp.bxp_brnfk
  WHERE  br.brn_name = ? AND bp.bxp_inactive = 0
    AND  bar.bar_deleted = 0 AND p.prd_deleted = 0
"""

BRANCHES_SQL = "SELECT brn_pk, brn_number, brn_name FROM Branches WHERE brn_Active = 1 ORDER BY brn_name"

def connect():
    for d in DRIVERS:
        try:
            return pyodbc.connect(f'DRIVER={{{d}}};SERVER={SERVER};DATABASE={DATABASE};Trusted_Connection=yes;TrustServerCertificate=yes;',timeout=5)
        except: continue
    raise Exception("Could not connect. Download ODBC driver from https://aka.ms/downloadmsodbcsql")

class H(http.server.BaseHTTPRequestHandler):
    def log_message(self,f,*a): print(f"[{datetime.now().strftime('%H:%M:%S')}] {f%a}")
    def send_json(self,code,data):
        b=json.dumps(data).encode()
        self.send_response(code)
        self.send_header('Content-Type','application/json')
        self.send_header('Access-Control-Allow-Origin','*')
        self.send_header('Content-Length',len(b))
        self.end_headers(); self.wfile.write(b)
    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header('Access-Control-Allow-Origin','*')
        self.send_header('Access-Control-Allow-Methods','GET,OPTIONS')
        self.end_headers()
    def do_GET(self):
        parsed=urllib.parse.urlparse(self.path)
        params=urllib.parse.parse_qs(parsed.query)
        if parsed.path=='/health':
            self.send_json(200,{'status':'ok'}); return
        if parsed.path=='/branches':
            try:
                c=connect(); cur=c.cursor(); cur.execute(BRANCHES_SQL)
                rows=[{'brn_pk':r[0],'brn_number':r[1],'brn_name':r[2]} for r in cur.fetchall()]
                c.close(); self.send_json(200,rows)
            except Exception as e: self.send_json(500,{'error':str(e)})
            return
        if parsed.path=='/stock':
            store=(params.get('store',[''])[0])
            if not store: self.send_json(400,{'error':'Missing ?store='}); return
            try:
                c=connect(); cur=c.cursor()
                cur.execute(STOCK_SQL, store)
                rows=cur.fetchall()
                if not rows:
                    cur.execute(STOCK_SQL.replace('br.brn_name = ?','br.brn_name LIKE ?'), f'%{store}%')
                    rows=cur.fetchall()
                c.close()
                m={r[0].strip():float(r[1] or 0) for r in rows if r[0]}
                print(f"  -> {store}: {len(rows)} products")
                self.send_json(200,m)
            except Exception as e: self.send_json(500,{'error':str(e)})
            return
        self.send_json(404,{'error':'Try /health /branches /stock?store=Name'})

if __name__=='__main__':
    s=http.server.HTTPServer(('127.0.0.1',PORT),H)
    print(f"\n✅  Beauty Star Orders server running")
    print(f"    http://localhost:{PORT}/branches")
    print(f"    http://localhost:{PORT}/stock?store=StoreName\n")
    try: s.serve_forever()
    except KeyboardInterrupt: print('\nStopped.'); s.server_close()
